lib/sinon/match.js

```diff
@@ -221,12 +221,50 @@ match.hasOwn = createPropertyMatcher(function (actual, property) {
     return actual.hasOwnProperty(property);
 }, "hasOwn");
 
+match.array = match.typeOf("array");
+
+match.array.deepEquals = function (expectation) {
+    return match(function (actual) {
+        // Comparing lengths is the fastest way to spot a difference before iterating through every item
+        var sameLength = actual.length === expectation.length;
+        return sameLength && Array.prototype.every.call(actual, function (element, index) {
```

Comment:
Sometimes people end up creating iterables without those methods, so if we just used `actual.every` or other methods directly on `actual` they might fail due to the lack of them being implemented. However they would still work when called this way. An example for this would be `arguments`. In order to make it an array and call these methods on it we need to use `Array.prototype.slice.call` before, since it doesn't have the other methods implemented.

Also, I think this approach is more functional and explicit, but this second reason is just a matter of personal opinion.